File: src/main/scala/kmeans/sec/Cluster.scala
package kmeans.sec

import kmeans.utils._

/**
  * Cluster con un centroide y una lista de puntos asignados
  * @param center Centroide del cluster
  * @param members Lista de puntos del cluster
  */

class Cluster(val center:CPoint,val members:List[CPoint]) {

  /**
    * Crea un nuevo cluster cuyo centroide es el punto medio de los puntos pertenecientes al anterior cluster
    * @return Cluster cuyo centroide es el punto medio de los puntos pertenecientes al cluster
    */

  def moveCenter:Cluster = {
    val point:CPoint = members
      .transpose // Obtenemos la traspuesta de la matriz
      .map(_.sum) // Para cada fila obtenemos la suma de sus valores
      .map(_/members.size) // Dividiendo entre el numero de puntos obtenemos el centroide
    Cluster(point)
  }

  /**
    * Calcula la distancia media de los puntos al centroide
    * @return Distancia media de los puntos al centroide
    */

  def avgDistanceCentroid:Double = {
    members
      .map(euclidean(center,_)) // Para cada punto obtenemos la distancia al centroide
      .sum/members.size // Obtenemos la distancia media
  }

  /**
    * Calcula la distancia de un punto al centroide
    * @param x Punto al que se le calculará la distancia
    * @return Distancia del punto x al centroide
    */

  def distanceToCentroid(x:CPoint):Double = euclidean(center,x)

  /**
    * Compara si dos clusters son iguales
    * @param obj Cluster con el que comparar
    * @return Comprobacion de si dos clusters son iguales
    */

  override def equals(obj: Any): Boolean = obj match {
    case (obj:Cluster) =>
      this.center == obj.center && this.members == obj.members
    case _ => false
  }
}

/** Factory Object para crear [[Cluster]] objetos */

object Cluster {

  /**
    * Crea un nuevo cluster dado un centroide
    * @param center Centroide del cluster
    * @return Cluster con un centroide asignado
    */

  def apply(center:CPoint) = new Cluster(center,Nil)

  /**
    * Crea un nuevo cluster dado un centroide y una lista de puntos
    * @param center Centroide del cluster
    * @param members Lista de puntos
    * @return Cluster con un centroide y una lista de puntos asignados
    */

  def apply(center:CPoint,
            members:List[CPoint])  = new Cluster(center,members)
}

File: src/main/java/sec/Cluster.java
package sec;

import java.util.ArrayList;

/**
 * Clase Cluster con porgramacion imperativa
 */

public class Cluster {

    /**
     * Centroide del Cluster
     */

    private Point center;

    /**
     * Lista de puntos pertenecientes al Cluster
     */

    private ArrayList<Point> members;

    /**
     * Constructor del Cluster
     * @param point Centroide del Cluster
     */

    public Cluster(Point point){
        center=new Point(point.getValues());
        members=new ArrayList<Point>();
    }

    /**
     * Contructor del Cluster
     * @param pt Centroide del Cluster
     * @param mbrs Lista de puntos
     */

    public Cluster(Point pt,ArrayList<Point> mbrs){
        center= new Point(pt.getValues());
        members = new ArrayList<Point>(mbrs);
    }

    /**
     * Getter del centroide del Cluster
     * @return Centroide del Cluster
     */

    public Point getCenter() {
        return center;
    }

    /**
     * Getter de la lista de puntos pertenecientes al Cluster
     * @return Lista de puntos
     */

    public ArrayList<Point> getMembers() {
        return members;
    }

    /**
     * Crea un nuevo cluster cuyo centroide es el punto medio
     * de los puntos pertenecientes al anterior cluster
     * @return Cluster cuyo centroide es el punto medio
     * de los puntos pertenecientes al cluster
     */

    public Cluster moveCluster(){
        if(members.size()==0)
            return this;

        ArrayList<Double> mean=new ArrayList<Double>();

        for(int i=0;i<center.getValues().size();++i){
            double m= 0.0;
            for(Point p:members)
                m+=(p.getValues().get(i));

            mean.add(m/members.size());
        }

        return new Cluster(new Point(mean));
    }

    /**
     * Calcula la distancia de un punto al centroide
     * @param p Punto al que se le calculará la distancia
     * @return Distancia del punto p al centroide
     * @throws Exception Si los puntos tienen diferente dimension
     */

    public double distanceToCentroid(Point p) throws Exception {
        return p.euclidean(center);
    }

    /**
     * Calcula la distancia media de los puntos al centroide
     * @return Distancia media de los puntos al centroide
     */

    public double avgDistanceToCentroid() throws Exception {
        double dist=0.0;
        for(Point p:members)
            dist+=p.euclidean(center);
        return dist/members.size();
    }

    /**
     * Compara si dos clusters son iguales
     * @param obj Cluster con el que comparar
     * @return Comprobacion de si dos clusters son iguales
     */

    @Override
    public boolean equals(Object obj){
        if( obj instanceof Cluster){
            if( center.equals(((Cluster) obj).getCenter())
                    && members.equals(((Cluster) obj).getMembers()))
                return true;
            else
                return false;
        }
        else
            return false;
    }
}
